File: src/main/java/dev/ambryn/task/services/TaskRunnerService.java
package dev.ambryn.task.services;

import dev.ambryn.task.models.Finished;
import dev.ambryn.task.models.Running;
import dev.ambryn.task.models.Task;
import dev.ambryn.task.utils.MathUtils;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.CancellationException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class TaskRunnerService {
    public static final int MAX_CONCURRENT_TASKS = 4;
    private final Queue<Task> taskQueue = new ArrayDeque<>();
    private final Map<Integer, Task> tasks = new HashMap<>();
    private final Map<Integer, CompletableFuture<Long>> taskHandles = new HashMap<>();
    private int currentRunningTasks = 0;

    public TaskRunnerService() {
        this.run();
    }

    public void addTasks(List<Task> tasks) {
        this.taskQueue.addAll(tasks);
        this.run();
    }

    public Iterable<Task> getQueuedTasks() {
        return Collections.unmodifiableCollection(this.taskQueue);
    }

    public Iterable<Task> getTasks() {
        return Stream.of(this.tasks.values()
                                   .stream()
                                   .toList(),
                         this.taskQueue.stream()
                                       .toList())
                     .flatMap(Collection::stream)
                     .toList();
    }

    public Set<Task> getRunningTasks() {
        return this.tasks.values()
                         .stream()
                         .filter(task -> task.getState() instanceof Running)
                         .collect(Collectors.toSet());
    }

    public Iterable<Task> getFinishedTasks() {
        return this.tasks.values()
                         .stream()
                         .filter(task -> task.getState() instanceof Finished)
                         .toList();
    }

    public Task getTask(int id) {
        return this.tasks.get(id);
    }

    public void cancelTask(int id) {
        this.taskHandles.get(id)
                        .completeExceptionally(new CancellationException("Canceled by user"));
    }

    public int sizeOfQueue() {
        return this.taskQueue.size();
    }

    private void run() {
        while (this.currentRunningTasks < MAX_CONCURRENT_TASKS && this.taskQueue.size() > 0) {
            Task task = this.taskQueue.remove();
            this.tasks.put(task.getId(), task);
            CompletableFuture<Long> taskHandle = this.runTask(task);

            taskHandle.thenAccept((result) -> {
                task.setResult(result);
                task.finish();
                task.describe();
                this.currentRunningTasks--;
                this.run();
            });
            taskHandle.exceptionally((ex) -> {
                if (ex instanceof CancellationException) {
                    task.cancel();
                } else {
                    System.out.println(ex.getMessage());
                    task.error();
                }
                this.currentRunningTasks--;
                return null;
            });

            this.taskHandles.put(task.getId(), taskHandle);
        }
    }

    public void clearAll() {
        this.taskQueue.clear();
        this.tasks.clear();
        this.taskHandles.clear();
        this.currentRunningTasks = 0;
    }

    private CompletableFuture<Long> runTask(Task task) {
        task.start();
        this.currentRunningTasks++;
        return CompletableFuture.supplyAsync(() -> {
            try {
                return MathUtils.fib(task.getNth());
            } catch (CancellationException e) {
                throw new CompletionException(e);
            }
        });
    }
}

File: src/main/java/dev/ambryn/task/services/TaskRunnerService.kt
package dev.ambryn.task.services

import dev.ambryn.task.models.*
import dev.ambryn.task.utils.MathUtils
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.CancellationException
import java.util.stream.Collectors
import java.util.stream.Stream

@Service
class TaskRunnerService {
    private val taskQueue: Queue<Task> = ArrayDeque()
    private val tasks: MutableMap<Int, Task> = HashMap()
    private val taskHandles: MutableMap<Int, Job> = HashMap()
    private var currentRunningTasks = 0

    fun addTasks(tasks: List<Task>?) {
        taskQueue.addAll(tasks!!)
        this.run()
    }

    fun getTasks(): Collection<Task> {
        return Stream.of(
            tasks.values.stream().toList(), taskQueue.stream().toList()
        ).flatMap { obj: List<Task> -> obj.stream() }.collect(Collectors.toSet())
    }

    val queuedTasks: Collection<Task>
        get() = Collections.unmodifiableCollection(taskQueue)
    val runningTasks: Collection<Task>
        get() = tasks.values.stream().filter { task: Task -> task.state is Running }.collect(Collectors.toSet())
    val finishedTasks: Collection<Task>
        get() = tasks.values.stream().filter { task: Task -> task.state is Finished }.collect(Collectors.toSet())
    val erroredTasks: Collection<Task>
        get() = tasks.values.stream().filter { task: Task -> task.state is Errored }.collect(Collectors.toSet())
    val cancelledTasks: Collection<Task>
        get() = tasks.values.stream().filter { task: Task -> task.state is Cancelled }.collect(Collectors.toSet())

    fun getTask(id: Int): Task? {
        return tasks[id]
    }

    fun cancelTask(id: Int) {
        println("Canceling task id=${id}")
        tasks[id]?.cancel()
        taskQueue.removeIf { task: Task -> task.id == id }
        taskHandles[id]?.cancel(CancellationException("Task id=${id} has been canceled by user"))
    }

    fun sizeOfQueue(): Int {
        return taskQueue.size
    }

    private fun run() {
        while (currentRunningTasks < MAX_CONCURRENT_TASKS && taskQueue.size > 0) {
            val task = taskQueue.remove()
            tasks[task.id] = task

            val taskHandle: Job = GlobalScope.launch {
                runTaskAsync(task)
            }

            taskHandles[task.id] = taskHandle
        }
    }

    fun clearAll() {
        taskQueue.clear()
        tasks.clear()
        taskHandles.clear()
        currentRunningTasks = 0
    }

    private suspend fun runTaskAsync(task: Task): Unit {
        task.start()
        currentRunningTasks++

        task.result = MathUtils.fib(task.nth)
        task.finish()
        task.describe()
        currentRunningTasks--

    }

    companion object {
        const val MAX_CONCURRENT_TASKS = 4
    }
}
